feat(submodules): use env vars from getGitEnvironmentVariables

RGunning opened the pull request:
<!-- If this is your first pull request: sign the CLA with this GitHub app: https://cla-assistant.io/renovatebot/renovate -->
<!-- Make sure the `Allow edits and access to secrets by maintainers` checkbox is checked on this pull request. -->
<!-- Please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md before you create your pull request.-->

Replacement PR for #31934 with imports fixed

## Changes

This change uses the existing getGitEnvironmentVariables function during the cloning of git submodules.

This function has been used to include support for updating private git ssh submodules on github.com, but it was not added in the function triggered when the cloneSubmodules config file option is set to true.

see: https://github.com/renovatebot/renovate/pull/22061

## Context

https://github.com/renovatebot/renovate/discussions/30942

Private submodules on the github.com app cannot currently be cloned with the cloneSubmodules option when the .gitmodules file is configured with ssh.

This is because the GitHub App token does not have SSH permissions, it can only auth via HTTP.

My use case is private Go submodules on github.com, for the "postUpdateOptions": ["gomodTidy", "gomodUpdateImportPaths"], options to work, I need my private git submodules cloned.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/renovatebot/renovate?pullRequest=35180) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/renovatebot/renovate?pullRequest=35180) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: tim-thacker-nullify<br/>:white_check_mark: rarkins<br/>:x: rsahrakorpi-cambri<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/renovatebot/renovate?pullRequest=35180) it.</sub>

viceice commented on lib/util/git/index.ts:
```diff
@@ -358,6 +359,13 @@ export async function cloneSubmodules(
+  const gitSubmoduleAuthEnvironmentVariables = getGitEnvironmentVariables();
+  const gitEnv = {
+    // pass all existing env variables
+    ...process.env,
```
@rarkins I would prefer to use `getChildProcessEnv` for security reasons.

viceice replied:
- #35764

viceice replied:
@rarkins use `getEnv` or `getChildEnv` here? 🤔

rarkins replied:
Not sure. `getEnv()` maybe?

viceice replied:
i think `getChildEnv`should be used here:
https://github.com/renovatebot/renovate/blob/bb9faed9c3754c316bea5415a3100df4284ce639/lib/modules/datasource/git-refs/base.ts#L37

RGunning replied:
Updated to use getChildEnv and confirmed that it still works after the change

RGunning commented:
Also, it appears the previous workaround we were using to get this working with gitSubmodules (https://qiita.com/sijiaoh/items/7d31904a4329d644b05d) stopped working due to https://github.com/renovatebot/renovate/pull/35764, so we are now blocked from using renovate until this PR is merged

rarkins commented:
Needs a test fix

viceice reviewed (commented):
needs test fix

RGunning committed:
fix(git): enhance submodule cloning with environment variables

RGunning commented:
I'm unable to run these tests locally, so can't fix the failing test. Trying to run locally I encounter a whole host of git issues that block the tests running in this branch or on main. Is there some docker environment where I can run the tests?

![Screenshot 2025-05-26 at 8 31 15](https://github.com/user-attachments/assets/b8ff2921-057e-4001-9dad-0662ee4c6f7c)

RGunning commented:
I'm unable to run these tests locally even in the devContainer. I'm getting a git error. Is there some documentation on how to setup and run these tests?

![Screenshot 2025-05-26 at 8 59 11](https://github.com/user-attachments/assets/bdce7f40-f192-4cfd-a438-86d1d800e3b0)

viceice reviewed (commented):
otherwise LGTM

RGunning committed:
refactor(git): inline getGitEnvironmentVariables into getChildEnv

Simplified the environment setup by inlining getGitEnvironmentVariables
directly into getChildEnv instead of assigning it to a separate constant.
This reduces indirection and improves readability.

viceice commented on lib/util/git/index.ts:
```diff
@@ -1,10 +1,10 @@
-import URL from 'node:url';
```
lint issue: import order is wrong

rsahrakorpi-cambri replied:
import order fixed, but tests still fail (but they also fail for me on main, so no idea what is wrong with them)

RGunning committed:
fix(git): import order

rarkins commented:
The failing test is in https://github.com/renovatebot/renovate/actions/runs/15993362696/job/45111113890?pr=35180

RGunning committed:
fix(tests): mock modifiedAuth and update submodule test case

RGunning commented on lib/util/git/index.spec.ts:
```diff
@@ -208,6 +211,10 @@ describe('util/git/index', { timeout: 10000 }, () => {
         const repo = Git(base.path);
 
+        modifiedAuth.getGitEnvironmentVariables.mockReturnValue({
+          GIT_ALLOW_PROTOCOL: 'file',
```
This was the minimal git config required for the tests to pass. Otherwise, this error is thrown 
> Error: fatal: not a git repository: ../.git/modules/submodule
fatal: could not reset submodule index

without this `getGitEnvironmentVariables` returns {}, which messes up the tests. This only seems to be required when running the unit tests, I guess, when running the code for real git inherits the required environment, or there are always HOST rules added, which might be sufficient to inform git that this is actually a git repository. I'm not sure on the underlying requirements of git

RGunning committed:
Merge branch 'main' into clone-private-submodules-support

viceice commented on lib/util/git/index.spec.ts:
```diff
@@ -10,24 +10,27 @@ import {
 import type { FileChange } from './types';
-import * as git from '.';
 import { setNoVerify } from '.';
+import * as git from '.';
```
please avoid this change, drop the line above this line and prefix that function call with `git.`

viceice commented on lib/util/git/index.spec.ts:
```diff
@@ -10,24 +10,27 @@ import {
   UNKNOWN_ERROR,
 } from '../../constants/error-messages';
 import { newlineRegex, regEx } from '../regex';
+import * as _modifiedAuth from './auth';
```
```suggestion
import * as _auth from './auth';
```
please no unnecessary renaming

viceice commented on lib/util/git/index.spec.ts:
```diff
@@ -10,24 +10,27 @@ import {
 const behindBaseCache = vi.mocked(_behindBaseCache);
 const conflictsCache = vi.mocked(_conflictsCache);
 const modifiedCache = vi.mocked(_modifiedCache);
+const modifiedAuth = vi.mocked(_modifiedAuth);
```
```suggestion
const auth = vi.mocked(_auth);
```

viceice commented on lib/util/git/index.spec.ts:
```diff
@@ -209,6 +212,10 @@ describe('util/git/index', { timeout: 10000 }, () => {
       beforeEach(async () => {
         const repo = Git(base.path);
 
+        modifiedAuth.getGitEnvironmentVariables.mockReturnValue({
```
```suggestion
        auth.getGitEnvironmentVariables.mockReturnValue({
```

viceice reviewed (commented):
otherwise LGTM

RGunning committed:
fix(tests): rename modifiedAuth to auth in git index tests

RGunning committed:
Merge branch 'clone-private-submodules-support' of personalgit:RGunning/renovate into clone-private-submodules-support

secustor reviewed (approved):
Theoretically breaking if users inject different credentials for submodule clones via config file, but I see this rather as a fix as it aligns the behaviour.
